PARQUET-2485: Be more consistent with BYTE_ARRAY types (#251)
apache/parquet-format: LogicalTypes.md

### STRING

> `STRING` may only be used to annotate the BYTE_ARRAY primitive type and indicates

Reviewer: It seems the spec is unclear about whether or not `STRING` and `ENUM` can annotate `FIXED_LENGTH_BYTE_ARRAY`. Literally it is reasonable to annotate `FIXED_LENGTH_BYTE_ARRAY`, right? I'm not sure if there is any use case in the wild.

Author: I don't know. For `DECIMAL` the spec calls out "byte arrays, binary and fixed" as valid physical types. I'd take the lack of mention of fixed-length here to indicated that only the `BYTE_ARRAY` physical type is allowed. Do any current implementations allow fixed-length strings?

Reviewer 2: I think the existing wording is pretty clear of we take "binary primitive type" to mean BYTE_ARRAY and thus in my opinion this is not a change to the spec. 

Perhaps we could send a note to dev@parquet.apache.org just highlighting this clarification in case someone wants to chime in and say they read it to mean FIXED_LENGTH_BYTE_ARRAY was also supported

Author: The sense I'm getting from the M/L is that people are open to adding it, but haven't seen any evidence that any one actually supports adding `STRING` to `FLBA`. IMO we should ship this as is, and start a new thread to gauge support for actually modifying the spec to allow for fixed length strings. Given the variable width of UTF-8 characters, I'd think padding would have to be added to account for up to 4 bytes per character.

Reviewer 2: I agree with this assesment

Reviewer: I agree that we'd better keep it as is for now. Let me merge this. We can always change the content when there is a consensus.

`DECIMAL` annotation represents arbitrary-precision signed decimal numbers of
the form `unscaledValue * 10^(-scale)`.

> The primitive type stores an unscaled integer value. For byte arrays, variable
> and fixed-length, the unscaled number must be encoded as two's complement using

Reviewer 2: We can probably make this wording clearer by just using the names directly: 

```suggestion
The primitive type stores an unscaled integer value. For BYTE_ARRAY and 
FIXED_LEN_BYTE_ARRAY, the unscaled number must be encoded as two's complement using
```

  warning
* `fixed_len_byte_array`: precision is limited by the array size. Length `n`
  can store &lt;= `floor(log_10(2^(8*n - 1) - 1))` base-10 digits

> * `byte_array`: `precision` is not limited, but is required. The minimum number of

Reviewer 2: Perhaps we can change this to name the types using the same names as elswhere:

```suggestion
* `FIXED_LEN_BYTE_ARRAY `: precision is limited by the array size. Length `n`
  can store &lt;= `floor(log_10(2^(8*n - 1) - 1))` base-10 digits
* `BYTE_ARRAY `: `precision` is not limited, but is required. The minimum number of

Author: Yes, the formatting here is odd...all of the type names are lower case. Also, the use of back ticks for type names is inconsistent.  Perhaps (as you suggest) for this PR we can solely worry about binary->byte_array, and then do a second pass to fix capitalization and quoting.

Author: Looking at this more, it seems I'm introducing more inconsistency with the quoting. I'll try to clean that up.

### JSON

`JSON` is used for an embedded JSON document. It must annotate a BYTE_ARRAY

> primitive type. The byte array data is interpreted as a UTF-8 encoded character

Reviewer 2: ```suggestion
primitive type. The BYTE_ARRAY data is interpreted as a UTF-8 encoded character
```

### BSON

`BSON` is used for an embedded BSON document. It must annotate a BYTE_ARRAY

> primitive type. The byte array data is interpreted as an encoded BSON document as

Reviewer 2: ```suggestion
primitive type. The BYTE_ARRAY data is interpreted as an encoded BSON document as
```
